invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+	// case the Lightning payment fails.
+	// Optional.
+	FallbackAddr btcutil.Address
+
+	// RoutingInfo is one or more entries containing extra routing
+	// information for a private route to the target node.
+	// Optional.
+	RoutingInfo []ExtraRoutingInfo
+}
+
+// ExtraRoutingInfo holds the information needed to route a payment along one
+// private channel.
+type ExtraRoutingInfo struct {
+	PubKey       *btcec.PublicKey
+	ShortChanID  uint64
+	Fee          uint64
+	CltvExpDelta uint16
+}
+
+// Amount is a functional option that allows callers of NewInvoice to set the
+// amount in millisatoshis that the Invoice should encode.
+func Amount(milliSat uint64) func(*Invoice) {
+	return func(i *Invoice) {
+		i.MilliSat = &milliSat
+	}
+}
+
+// PubKey is a functional option that allows callers of NewInvoice to explicitly
+// set the pubkey of the Invoice's target node.
+func PubKey(pubKey *btcec.PublicKey) func(*Invoice) {
+	return func(i *Invoice) {
+		i.PubKey = pubKey
+	}
+}
+
+// Description is a functional option that allows callers of NewInvoice to set
+// the payment description of the created Invoice.
+// Note: Must be used if and only if DescriptionHash is not used.
+func Description(description string) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Description = &description
+	}
+}
+
+// DescriptionHash is a functional option that allows callers of NewInvoice to
+// set the payment description hash of the created Invoice.
+// Note: Must be used if and only if Description is not used.
+func DescriptionHash(descriptionHash []byte) func(*Invoice) {
+	return func(i *Invoice) {
+		i.DescriptionHash = descriptionHash
+	}
+}
+
+// Expiry is a functional option that allows callers of NewInvoice to set the
+// expiry of the created Invoice. If not set, a default expiry of 60 min will
+// be implied.
+func Expiry(expiry uint64) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Expiry = &expiry
+	}
+}
+
+// FallbackAddr is a functional option that allows callers of NewInvoice to set
+// the Invoice's fallback on-chain address that can be used for payment in case
+// the Lightning payment fails
+func FallbackAddr(fallbackAddr btcutil.Address) func(*Invoice) {
+	return func(i *Invoice) {
+		i.FallbackAddr = fallbackAddr
+	}
+}
+
+// RoutingInfo is a functional option that allows callers of NewInvoice to set
+// one or more entries containing extra routing information for a private route
+// to the target node.
+func RoutingInfo(routingInfo []ExtraRoutingInfo) func(*Invoice) {
+	return func(i *Invoice) {
+		i.RoutingInfo = routingInfo
+	}
+}
+
+// NewInvoice creates a new Invoice object. The last parameter is a set of
+// variadic argumements for setting optional fields of the invoice.
+// Note: Either Description  or DescriptionHash must be provided for the Invoice
+// to be considered valid.
+func NewInvoice(net *chaincfg.Params, paymentHash []byte,
+	timestamp uint64, options ...func(*Invoice)) *Invoice {
```

Comment:
Perhaps we should also add a `Validate()` method and call that here before returning. The `Validate()` method should perform sanity checks of all the inputs ensuring for example that the number of passed routing info hints doesn't exceed 20. As a result, the function will now return an additional error.

Reply:
Getting a bit further in I see there's a `validateInvoice` function that can be used here.